packages/ecmascript/src/Module.ts

```diff
@@ -142,60 +249,110 @@ export interface ModuleElementList {
-				let circPath = loadStack.slice(indexOfPathInStack);
+				const circPath = loadStack.slice(indexOfPathInStack);
 				circPath.push(path);
 				loadStack = [];
-				throw new Error(`Detected circular dependency in module loading. Path: ${JSON.stringify(circPath)}`);
-			}
-
-			loadStack.push(path);
-			try {
-				actionResult = invokeActionOrModule(path);
+				error = `Detected circular dependency in module loading. Path: ${JSON.stringify(circPath)}`;
 			}
-			finally {
-				loadStack.pop();
+			else {
+				loadStack.push(path);
+				try {
+					actionResult = invokeActionOrModule(path, onError);
+				}
+				catch (err) {
+					error = err?.message || err;
+				}
+				finally {
+					loadStack.pop();
+				}
 			}
-
-			actionResults[path] = actionResult;
+			actionResults[path] = actionResult || null; // won't reattempt unsuccessful load
 		}
 
-		return actionResult;
+		return actionResult || onError(`Failed to load action or module with path '${path}'! ${error}`);
 	}
 
-	function invokeActionOrModule(path: string): any {
-		let classIndex = path.lastIndexOf(".");
-		let moduleName = path.substring(0, classIndex);
-		let actionName = path.substring(classIndex + 1);
-		let moduleInfo = <Module>System.getModule(moduleName);
-		if (moduleInfo && moduleInfo.actionDescriptions.some(a => a.name == actionName)) {
+	/**
+	 * Attempts to invoke the action or module (namespace) with the given path.
+	 * @param {string} path - absolute path with dot (.) separators. Cannot be blank or null/undefined.
+	 * @param {ErrorHandler} [onError = DEFAULT_ERR_HANDLER] - function to handle errors when loading a module/action.
+	 * Passed on recursively to a module's child modules - to be invoked on creation.
+	 * @returns any - action or module:
+	 * - If there is an action at the given path, it will be returned.
+	 * - if there is a module with an "index" action at the given path, the "index" action will be returned.
+	 * - if there is a module at the given path without an "index" action, a module descriptor for it will be returned.
+	 * @throws Error if no module exists for the given path.
+	 */
+	function invokeActionOrModule(path: string, onError: ErrorHandler = DEFAULT_ERR_HANDLER): any {
+		const classIndex = path.lastIndexOf(".");
+		const moduleName = path.substring(0, classIndex);
+		const actionName = path.substring(classIndex + 1);
+		let moduleInfo: Module = !moduleName ? null : System.getModule(moduleName);
+		if (hasAction(moduleInfo, actionName)) {
 			return invokeAction(moduleInfo, actionName);
 		}
-		else {
-			moduleInfo = <Module>System.getModule(path);
-			if (moduleInfo) {
-				if (moduleInfo.actionDescriptions.some(a => a.name == "index")) {
-					return invokeAction(moduleInfo, "index")
-				}
-				else {
-					return createModule(moduleInfo);
-				}
-			}
+		moduleInfo = !path ? null : System.getModule(path);
+		if (!moduleInfo) {
+			throw new Error(`No action or module found for paths: '${path}', '${path}/index'!`);
```

Comment:
Maybe we could include moduleName, actionName as additional information to error message. In case of typos, strange chars, etc it could be helpful.

Reply:
Path is included for that purpose. It is unknown whether an action or a module (namespace) is being imported.